diff -crB p4est/src/p4est.c p4est_update/src/p4est.c
*** p4est/src/p4est.c	2016-09-08 12:15:06.000000000 -0700
--- p4est_update/src/p4est.c	2018-05-25 19:29:06.000000000 -0700
***************
*** 201,206 ****
--- 201,219 ----
                 p4est_locidx_t min_quadrants, int min_level, int fill_uniform,
                 size_t data_size, p4est_init_t init_fn, void *user_pointer)
  {
+   p4est_t * p4est;
+   p4est = P4EST_ALLOC_ZERO (p4est_t, 1);
+   p4est_init_ext (p4est, mpicomm, connectivity, 0, 0, 1,
+                   data_size, init_fn, user_pointer);
+   return p4est;
+ }
+ 
+ void
+ p4est_init_ext (p4est_t * p4est, sc_MPI_Comm mpicomm,
+                 p4est_connectivity_t * connectivity,
+                 p4est_locidx_t min_quadrants, int min_level, int fill_uniform,
+                 size_t data_size, p4est_init_t init_fn, void *user_pointer)
+ {
    int                 num_procs, rank;
    int                 i, must_remove_last_quadrant;
    int                 level;
***************
*** 210,216 ****
    p4est_gloidx_t      first_tree, first_quadrant, first_tree_quadrant;
    p4est_gloidx_t      last_tree, last_quadrant, last_tree_quadrant;
    p4est_gloidx_t      quadrant_index;
-   p4est_t            *p4est;
    p4est_tree_t       *tree;
    p4est_quadrant_t   *quad;
    p4est_quadrant_t    a, b, c;
--- 223,228 ----
***************
*** 227,233 ****
    P4EST_ASSERT (min_level <= P4EST_QMAXLEVEL);
  
    /* create p4est object and assign some data members */
!   p4est = P4EST_ALLOC_ZERO (p4est_t, 1);
    p4est->data_size = data_size;
    p4est->user_pointer = user_pointer;
    p4est->connectivity = connectivity;
--- 239,245 ----
    P4EST_ASSERT (min_level <= P4EST_QMAXLEVEL);
  
    /* create p4est object and assign some data members */
!   memset(p4est, 0, sizeof(p4est_t));
    p4est->data_size = data_size;
    p4est->user_pointer = user_pointer;
    p4est->connectivity = connectivity;
***************
*** 491,502 ****
    P4EST_GLOBAL_PRODUCTIONF ("Done " P4EST_STRING
                              "_new with %lld total quadrants\n",
                              (long long) p4est->global_num_quadrants);
-   return p4est;
  }
  
  void
  p4est_destroy (p4est_t * p4est)
  {
  #ifdef P4EST_ENABLE_DEBUG
    size_t              qz;
  #endif
--- 503,519 ----
    P4EST_GLOBAL_PRODUCTIONF ("Done " P4EST_STRING
                              "_new with %lld total quadrants\n",
                              (long long) p4est->global_num_quadrants);
  }
  
  void
  p4est_destroy (p4est_t * p4est)
  {
+   p4est_destroy_ext (p4est, 0 == 0);
+ }
+ 
+ void
+ p4est_destroy_ext (p4est_t * p4est, int free_p4est)
+ {
  #ifdef P4EST_ENABLE_DEBUG
    size_t              qz;
  #endif
***************
*** 526,532 ****
    p4est_comm_parallel_env_release (p4est);
    P4EST_FREE (p4est->global_first_quadrant);
    P4EST_FREE (p4est->global_first_position);
!   P4EST_FREE (p4est);
  }
  
  p4est_t            *
--- 543,550 ----
    p4est_comm_parallel_env_release (p4est);
    P4EST_FREE (p4est->global_first_quadrant);
    P4EST_FREE (p4est->global_first_position);
!   if( free_p4est)
!     P4EST_FREE (p4est);
  }
  
  p4est_t            *
diff -crB p4est/src/p4est_extended.h p4est_update/src/p4est_extended.h
*** p4est/src/p4est_extended.h	2016-09-08 12:15:06.000000000 -0700
--- p4est_update/src/p4est_extended.h	2018-05-25 19:29:06.000000000 -0700
***************
*** 130,135 ****
--- 130,155 ----
                                     size_t data_size, p4est_init_t init_fn,
                                     void *user_pointer);
  
+ /** Initialize an already allocated forest
+  * See the documentation of p4est_new_ext for basic usage.
+  *
+  * \param [in] p4est  p4est to initialize
+  */
+ void                p4est_init_ext (p4est_t * p4est,
+                                     sc_MPI_Comm mpicomm,
+                                     p4est_connectivity_t * connectivity,
+                                     p4est_locidx_t min_quadrants,
+                                     int min_level, int fill_uniform,
+                                     size_t data_size, p4est_init_t init_fn,
+                                     void *user_pointer);
+ 
+ /** Destroy a p4est.
+  * See the documentation of p4est_destroy for basic usage.
+  *
+  * \param [in] destroy_p4est  bool on whether to free p4est
+  */
+ void                p4est_destroy_ext (p4est_t * p4est, int free_p4est);
+ 
  /** Create a new mesh.
   * \param [in] p4est                A forest that is fully 2:1 balanced.
   * \param [in] ghost                The ghost layer created from the
diff -crB p4est/src/p4est_to_p8est.h p4est_update/src/p4est_to_p8est.h
*** p4est/src/p4est_to_p8est.h	2016-09-08 12:15:06.000000000 -0700
--- p4est_update/src/p4est_to_p8est.h	2018-05-25 19:29:07.000000000 -0700
***************
*** 178,183 ****
--- 178,185 ----
  /* functions in p4est_extended */
  #define p4est_replace_t                 p8est_replace_t
  #define p4est_new_ext                   p8est_new_ext
+ #define p4est_init_ext                  p8est_init_ext
+ #define p4est_destroy_ext               p8est_destroy_ext
  #define p4est_mesh_new_ext              p8est_mesh_new_ext
  #define p4est_copy_ext                  p8est_copy_ext
  #define p4est_refine_ext                p8est_refine_ext
diff -crB p4est/src/p8est_extended.h p4est_update/src/p8est_extended.h
*** p4est/src/p8est_extended.h	2016-09-08 12:15:06.000000000 -0700
--- p4est_update/src/p8est_extended.h	2018-05-25 19:29:07.000000000 -0700
***************
*** 130,135 ****
--- 130,155 ----
                                     size_t data_size, p8est_init_t init_fn,
                                     void *user_pointer);
  
+ /** Initialize an already allocated forest
+  * See the documentation of p8est_new_ext for basic usage.
+  *
+  * \param [in] p8est  p8est to initialize
+  */
+ void                p8est_init_ext (p8est_t * p8est,
+                                     sc_MPI_Comm mpicomm,
+                                     p8est_connectivity_t * connectivity,
+                                     p4est_locidx_t min_quadrants,
+                                     int min_level, int fill_uniform,
+                                     size_t data_size, p8est_init_t init_fn,
+                                     void *user_pointer);
+ 
+ /** Destroy a p8est.
+  * See the documentation of p8est_destroy for basic usage.
+  *
+  * \param [in] destroy_p8est  bool on whether to free p8est
+  */
+ void                p8est_destroy_ext (p8est_t * p8est, int free_p8est);
+ 
  /** Create a new mesh.
   * \param [in] p8est                A forest that is fully 2:1 balanced.
   * \param [in] ghost                The ghost layer created from the
